ENH Preserving dtype for np.float32 in *DictionaryLearning, SparseCoder and orthogonal_mp_gram

takoika opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->

This PR is part of #11000 .

#### What does this implement/fix? Explain your changes.

This PR makes an obtained code and dictionary of Dictionary Learning numpy.float32 when input data is numpy.float32 in order to preserve input data type.

#### Any other comments?

I found two difficulties in testing numerical consistency between numpy.float32 and numpy.float64 for dictionary learning.
* optimal code and dictionary are not unique.
* difficult to match with high precision(maybe due to multiple linear algebra)

In the scope of the PR it is OK. But potentially this makes it difficult to guarantee numerical consistency for downstream methods to use Dictionary learning.

Further test cases may be required because this PR does not cover argument variations.

I used #13303, #13243 and #20155 as references to make this.
<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Add test for checking consistency beteen input data and learned dictionary

Commit:
Merge branch 'main' of github.com:takoika/scikit-learn into issue11000dict_learning

Commit:
Accept np.float32 for check_array

Commit:
Split test function for dict_learning and dict_learning_online

Commit:
Merge branch 'main' of github.com:takoika/scikit-learn into issue11000dict_learning

Commit:
Add unit tests for numerical consistency of dict learning between np.float32 and np.float64

Commit:
refine comments

ogrisel reviewed (approved):
Just a few questions/suggestions for improvements below but the PR already LGTM as it is.

Please do not forget to document the enhancement in `doc/whats_new/v1.1.rst`.

Commit:
Add change log

takoika commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -654,6 +654,130 @@ def test_warning_default_transform_alpha(Estimator):
+    n_components = 6
+    # The larger alpha, the more sparsity,
+    # as a result likely to achieve tolerance error.
+    alpha = 4
```
Thanks!
The solution for the parameters seems degenerate, because all of elements of U is 0. So I will change the parameters.

takoika replied:
I have changed test parameters not to be degenerate solution. And add test to ensure that U has non zero elements at least 5% .

Commit:
Fix pr number

jeremiedbb commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -654,6 +654,130 @@ def test_warning_default_transform_alpha(Estimator):
+    # and (column permutated U*, row permutated V*) are also optional
+    # as long as holding UV.
+    # So here UV, ||U||_1,1 and sum(||V_k||_2^2) are verified
+    # instead of comparing directory U and V.
```
```suggestion
    # instead of comparing directly U and V.
```

takoika replied:
Thx! Fixed

jeremiedbb commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -654,6 +654,130 @@ def test_warning_default_transform_alpha(Estimator):
+        X.astype(np.float64),
+        n_components=n_components,
+        alpha=alpha,
+        random_state=rng,
```
You should be using the same seed for both. Here rng is consumed between the 2 dict_learning calls and thus reproducibility is not ensured. Setting `random_state=0` in both will fix that

takoika replied:
Exactly! I have passed `rand_state=0` to the function.

jeremiedbb commented on doc/whats_new/v1.1.rst:
```diff
@@ -119,6 +119,9 @@ Changelog
   get accurate results when the number of features is large.
   :pr:`21109` by :user:`Smile <x-shadow-man>`.
 
+- |Enhancement| `dict_learning` and `dict_learning_online` methods preserve dtype for numpy.float32.
+  :pr:`22002` by :user:`Takeshi Oura <takoika>`.
+
```
Your changes also impact `DictionaryLearning`, `MiniBatchDictionaryLearning` and `SparseCoder`. Please mention them here. I think it would also be nice to add similar tests for those as well

takoika replied:
Changed.
Also unit tests to verify dtype matching for `DictionaryLearning`,` MiniBatchDictionaryLearning` and `SparseCoder` are added. Then a test for numerical consistency among `np.float32` and `np.float64` in `sparse_encode` method is added.

jeremiedbb commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -654,6 +654,130 @@ def test_warning_default_transform_alpha(Estimator):
+    # instead of comparing directory U and V.
+    assert_allclose(np.matmul(U_64, V_64), np.matmul(U_32, V_32), rtol=rtol, atol=atol)
+    assert_allclose(np.sum(np.abs(U_64)), np.sum(np.abs(U_32)), rtol=rtol, atol=atol)
+    assert_allclose(np.sum(V_64 ** 2), np.sum(V_32 ** 2), rtol=rtol, atol=atol)
```
I'd rather have only rtol than only atol. In addition I think rtol=1e-7 is a little bit too optimistic when comparing float32s because it's slightly lower than the machine precision. 1e-6 would be safer.

takoika replied:
Thanks!
I have changed to use only rtol. 
Anyway some tests are difficult to pass rtol=1e-7 or  1e-6. So  minimum rtol to pass tests are set.

Commit:
Update test parameters and add test for verifying not degenarate

Commit:
Fix typo

Commit:
Add tests for dtype preserving in sparse_encode, SparseCoder, DictionaryLearning and MiniBatchDictionaryLearning

Commit:
Add numerical consistency among np.float32 and np.float64 for sparse_encode

Commit:
Refactor for visibility

Commit:
update whats_new

Commit:
Relax rtol

jeremiedbb commented:
I suspect that we only need to change this to make "omp" preserve dtype as well
https://github.com/scikit-learn/scikit-learn/blob/532e1bc60819cea6d313aa8819a208d5a1e0d559/sklearn/linear_model/_omp.py#L557
https://github.com/scikit-learn/scikit-learn/blob/532e1bc60819cea6d313aa8819a208d5a1e0d559/sklearn/linear_model/_omp.py#L559
adding `dtype=Gram.dtype`.

Would you mind trying this ? If it requires more work, we can do it later in a separated PR.

Commit:
Initialize coef by same dtype with input gram matrix

Commit:
update changelog

takoika commented:
> I suspect that we only need to change this to make "omp" preserve dtype as well

Thank you for the suggestion.
By the proposed changes `omp` algorithm come to pass unit test. So I have added the change to this PR.

takoika commented:
@jeremiedbb 
The change of `scikit-learn/sklearn/linear_model/_omp.py`  changes the behaviour of `orthogonal_mp_gram`. So is it needed to mention it in change log?

Anyway `othogonal_mp`, `OrthogonalMatchingPursuit` and `OrthogonalMatchingPursuitCV` will be affected by the change as well as `orthogonal_mp_gram`. It would be better to add unit tests to confirm relations between input dtype and output dtype. But in terms of making PR small and simple, it would be better to address `_omp.py` at other PR.
Any thoughts?

glemaitre commented:
Something that will be required as well for the common test is to add the proper tag to the classes that are preserving the dtype. It boils down to adding the `"preserves_dtype"` key into the dictionary returned by `_more_tags`.

```python
def _more_tags(self):
        return {
            "preserves_dtype": [np.float64, np.float32],
        }
```

This should be added to:

- `DictionaryLearning`
- `MiniBatchDictionaryLearning`
- `SparseCoder`

Some checks are probably a bit duplicated with the one written but I don't think that this is a big deal because I am not sure that we are running the common test for all the above classes.

glemaitre commented:
> Anyway othogonal_mp, OrthogonalMatchingPursuit and OrthogonalMatchingPursuitCV will be affected by the change as well as orthogonal_mp_gram. It would be better to add unit tests to confirm relations between input dtype and output dtype for them. But in terms of making PR small and simple, it would be better to address _omp.py at next PR.
Any thoughts?

I would advise making it in a separate PR if this is not as straightforward regarding the behaviour.

Commit:
Remove incorrect todo

Commit:
Add test for preverving input dtype np.float32 in orthogonal_mp_gram

Commit:
Add changelog

Commit:
refactor

Commit:
Add preserves_dtype tag

takoika commented:
@glemaitre
Thanks!
> Something that will be required as well for the common test is to add the proper tag to the classes that are preserving the dtype. It boils down to adding the "preserves_dtype" key into the dictionary returned by _more_tags.

I have added `preserves_dtype` tag.

> I would advise making it in a separate PR if this is not as straightforward regarding the behaviour.

I have confirmed that the change affect only `orthogonal_mp_gram` . So I have added  unit tests to verify the behaviour that input `np.float32` is preserved as output `np.float32`.   `othogonal_mp`, `OrthogonalMatchingPursuit` and `OrthogonalMatchingPursuitCV` do not yet preserve `np.float32`. So I will leave as is.

glemaitre reviewed (approved):
LGTM on my side. @jeremiedbb are you OK with the PR as-is.

Commit:
Merge branch 'main' of github.com:takoika/scikit-learn into issue11000dict_learning

Commit:
Revert accidental change introduced by merging upstream

jeremiedbb reviewed (approved):
LGTM. Thanks @takoika !
